File: Complementar - Estruturas de dados (Didático)/Algoritmos12/Alg12.php
<?php
/*
♦ Algoritmo 12: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Ligada/Encadeada Simples (não Duplamente Ligada) de tipo de dados genérico.
→ Operações:
    • addList: inserção em uma Lista Ligada Simples;
    • removeList: remoção em uma Lista Ligada Simples;
    • size - obter o tamanho atual da Lista Ligada Simples;
    • showList: exibição dos elementos da Lista Ligada Simples.
*/

class Nodo {
    public $valor;
    public $proximo;

    public function __construct($valor) {
        $this->valor = $valor;
        $this->proximo = null;
    }
}

class ListaLigada {
    private $primeiro;
    private $tamanho;

    public function __construct() {
        $this->primeiro = null;
        $this->tamanho = 0;
    }

    public function addList($valor) {
        $novoNodo = new Nodo($valor);
        $novoNodo->proximo = $this->primeiro;
        $this->primeiro = $novoNodo;
        $this->tamanho++;
    }

    public function removeList($valor) {
        if ($this->primeiro === null) {
            return;
        }

        if ($this->primeiro->valor == $valor) {
            $this->primeiro = $this->primeiro->proximo;
            $this->tamanho--;
            return;
        }

        $nodoAtual = $this->primeiro;
        while ($nodoAtual->proximo !== null) {
            if ($nodoAtual->proximo->valor == $valor) {
                $nodoAtual->proximo = $nodoAtual->proximo->proximo;
                $this->tamanho--;
                return;
            }
            $nodoAtual = $nodoAtual->proximo;
        }
    }

    public function size() {
        return $this->tamanho;
    }

    public function showList() {
        $nodoAtual = $this->primeiro;
        while ($nodoAtual !== null) {
            echo $nodoAtual->valor . " ";
            $nodoAtual = $nodoAtual->proximo;
        }
        echo "\n";
    }
}

$lista = new ListaLigada();
$lista->addList(10);
$lista->addList(20);
$lista->addList(30);

echo "Tamanho da lista: " . $lista->size() . "\n";
echo "Elementos da lista: ";
$lista->showList();

$lista->removeList(20);

echo "Tamanho da lista após remoção: " . $lista->size() . "\n";
echo "Elementos da lista após remoção: ";
$lista->showList();

File: Complementar - Estruturas de dados (Didático)/Algoritmos12/Alg12.py
'''
♦ Algoritmo 12: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Ligada/Encadeada Simples (não Duplamente Ligada) de tipo de dados genérico.
→ Operações:
    • addList: inserção em uma Lista Ligada Simples;
    • removeList: remoção em uma Lista Ligada Simples;
    • size - obter o tamanho atual da Lista Ligada Simples;
    • showList: exibição dos elementos da Lista Ligada Simples.
'''

class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None

class ListaLigada:
    def __init__(self):
        self.primeiro = None
        self.tamanho = 0

    def addList(self, valor):
        novoNodo = Nodo(valor)
        novoNodo.proximo = self.primeiro
        self.primeiro = novoNodo
        self.tamanho += 1

    def removeList(self, valor):
        if self.primeiro is None:
            return

        if self.primeiro.valor == valor:
            self.primeiro = self.primeiro.proximo
            self.tamanho -= 1
            return

        nodoAtual = self.primeiro
        while nodoAtual.proximo is not None:
            if nodoAtual.proximo.valor == valor:
                nodoAtual.proximo = nodoAtual.proximo.proximo
                self.tamanho -= 1
                return
            nodoAtual = nodoAtual.proximo

    def size(self):
        return self.tamanho

    def showList(self):
        nodoAtual = self.primeiro
        while nodoAtual is not None:
            print(nodoAtual.valor, end=" ")
            nodoAtual = nodoAtual.proximo
        print()

lista = ListaLigada()
lista.addList(10)
lista.addList(20)
lista.addList(30)

print("Tamanho da lista:", lista.size())
print("Elementos da lista:", end=" ")
lista.showList()

lista.removeList(20)

print("Tamanho da lista após remoção:", lista.size())
print("Elementos da lista após remoção:", end=" ")
lista.showList()
